web3/account.py

```diff
@@ -0,0 +1,189 @@
+        pubkey = signature.recover_public_key_from_msg_hash(hash_bytes)
+        return pubkey.to_checksum_address()
+
+    def recoverTransaction(self, primitive=None, hexstr=None):
+        raw_tx = to_bytes(primitive, hexstr=hexstr)
+        # TODO move to a new transaction handler
+        tx_parts = rlp.decode(raw_tx)
+        unsigned_parts = tx_parts[:-3]
+        raw_v, r, s = map(to_decimal, tx_parts[-3:])
+        (chain_aware_tx, _chain_id, v) = annotate_transaction_with_chain_id(unsigned_parts, raw_v)
+        message_hash = keccak(rlp.encode(chain_aware_tx))
+        return self.recover(message_hash, vrs=(v, r, s))
+
+    def setKeyBackend(self, backend):
+        self._keys = KeyAPI(backend)
+
+    def sign(self, message=None, private_key=None, message_hexstr=None, message_text=None):
+        '''
+        @param private_key in bytes, str, or int.
+            In Python 2, a bytes, unicode or str object will be interpreted as hexstr
+            In Python 3, only a str object will be interpreted as hexstr
+        '''
+        msg_bytes = to_bytes(message, hexstr=message_hexstr, text=message_text)
+        msg_hash = self.hashMessage(msg_bytes)
+        if isinstance(private_key, str) or (
+                sys.version_info.major < 3 and isinstance(private_key, unicode)  # noqa: F821
+            ):
+            key_bytes = to_bytes(hexstr=private_key)
+        else:
+            key_bytes = to_bytes(private_key)
+        key = self._keys.PrivateKey(key_bytes)
+        # TODO move to a new signature handler
+        signature = key.sign_msg_hash(to_bytes(hexstr=msg_hash))
+        (r, s, v_raw) = (getattr(signature, part) for part in 'rsv')
+        v = v_raw + 27
+        eth_signature_bytes = b''.join(map(to_bytes, (r, s, v)))
+        (r_hex, s_hex, eth_signature_hex) = map(to_hex, (r, s, eth_signature_bytes))
+        return AttributeDict({
+            'message': msg_bytes,
+            'messageHash': msg_hash,
+            'r': r_hex,
+            's': s_hex,
+            'v': v,
+            'signature': eth_signature_hex,
+        })
+
+    def signTransaction(self, transaction_dict, private_key):
+        '''
+        @param private_key in bytes, str, or int.
+            In Python 2, a bytes, unicode or str object will be interpreted as hexstr
+            In Python 3, only a str object will be interpreted as hexstr
+        '''
+        assert isinstance(transaction_dict, Mapping)
+
+        # build private key
+        if isinstance(private_key, str) or (
+                sys.version_info.major < 3 and isinstance(private_key, unicode)  # noqa: F821
+            ):
+            key_bytes = to_bytes(hexstr=private_key)
+        else:
+            key_bytes = to_bytes(private_key)
+        account = self.privateKeyToAccount(key_bytes)
+
+        # detect nonce for this account
+        if 'nonce' not in transaction_dict:
+            transaction_dict['nonce'] = self.web3.eth.getTransactionCount(account.address)
```

Comment:
This is an acceptable solution for initial implementation but it will fail anytime two transactions are signed before the 1st transaction has been mined since `getTransactionCount` will return the same value.

Eventually we will need to perform local tracking on the nonce value, taking into account any transactions we've signed locally and broadcast.

Reply:
Yup, agreed. It looks like web3.js is punting on internal nonce management too... at least according to their docs.